va/http.go

```diff
@@ -0,0 +1,424 @@
+package va
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"net/url"
+	"strconv"
+	"time"
+
+	"github.com/letsencrypt/boulder/core"
+	berrors "github.com/letsencrypt/boulder/errors"
+	"github.com/letsencrypt/boulder/probs"
+)
+
+func newHTTPClient() http.Client {
+	// Construct a one-off HTTP client with a custom transport.
+	return http.Client{
+		Transport: &http.Transport{
+			// We are talking to a client that does not yet have a certificate,
+			// so we accept a temporary, invalid one.
+			TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+			// We don't expect to make multiple requests to a client, so close
+			// connection immediately.
+			DisableKeepAlives: true,
+			// We don't want idle connections, but 0 means "unlimited," so we pick 1.
+			MaxIdleConns:        1,
+			IdleConnTimeout:     time.Second,
+			TLSHandshakeTimeout: 10 * time.Second,
+		},
+	}
+}
+
+// httpValidationURL constructs a URL for the given IP address, path and port
+// combination. The port is omittied from the URL if it is the default HTTP
+// port. It is only used for constructing initial HTTP-01 validation requests
+// and so does not need to support constructing a URL with an HTTPS scheme.
+func httpValidationURL(validationIP net.IP, path string, port int) *url.URL {
+	urlHost := validationIP.String()
+
+	// If the port is something other than the conventional HTTP port, put it in
+	// the URL.
+	if port != 80 {
+		urlHost = net.JoinHostPort(validationIP.String(), strconv.Itoa(port))
+	} else if validationIP.To4() == nil {
+		// if the validation IP is an IPv6 address, and we aren't using
+		// `net.JoinHostPort` then we have to manually surround the IPv6 address
+		// with square brackets to make a valid IPv6 URL (e.g "http://[::1]/foo" not
+		// "http://::1/foo")
+		urlHost = fmt.Sprintf("[%s]", urlHost)
+	}
+
+	return &url.URL{
+		Scheme: "http",
+		Host:   urlHost,
+		Path:   path,
+	}
+}
+
+// httpValidationTarget bundles all of the information needed to make an HTTP-01
+// validation request against a target.
+type httpValidationTarget struct {
+	// the hostname being validated
+	host string
+	// the port for the validation request
+	port int
+	// the path for the validation request
+	path string
+	// all of the IP addresses available for the host
+	available []net.IP
+	// the IP addresses that were tried for validation so far (e.g. returned by
+	// calls to the IP function)
+	tried []net.IP
+	// the IP addresses that will be returned by calls to the IP function.
+	next []net.IP
+}
+
+// IP hands out one of the next IP addresses for the validation target. The list
+// of tried IP addresses is updated to include the IP that is returned to the
```

Comment:
This comment says that the list of tried IP addresses is updated, but `next` is also modified. Ideally this comment should make very clear that this method mutates the state of the `httpValidationTarget` it is called on.